bigtable/tests/filters_integration_test.cc

```diff
@@ -0,0 +1,693 @@
+  request.set_rows_limit(1);
+  auto& row = *request.mutable_rows();
+  *row.add_row_keys() = std::move(key);
+  *request.mutable_filter() = filter.as_proto_move();
+
+  return ReadRows(client, table, std::move(request));
+}
+
+// TODO(#32) remove this when Table::ReadRow() is a thing.
+std::vector<bigtable::Cell> ReadRows(bigtable::DataClient& client,
+                                     bigtable::Table& table, std::string start,
+                                     std::string end, bigtable::Filter filter) {
+  btproto::ReadRowsRequest request;
+  request.set_table_name(table.table_name());
+  auto& row = *request.mutable_rows();
+  btproto::RowRange range;
+  range.set_start_key_closed(std::move(start));
+  range.set_end_key_open(std::move(end));
+  *row.add_row_ranges() = std::move(range);
+  *request.mutable_filter() = filter.as_proto_move();
+  return ReadRows(client, table, std::move(request));
+}
+
+/// A helper function to create a list of cells.
+void CreateCells(bigtable::Table& table,
+                 std::vector<bigtable::Cell> const& cells) {
+  std::map<std::string, bigtable::SingleRowMutation> mutations;
+  for (auto const& cell : cells) {
+    std::string key = static_cast<std::string>(cell.row_key());
+    auto inserted = mutations.emplace(key, bigtable::SingleRowMutation(key));
+    inserted.first->second.emplace_back(bigtable::SetCell(
+        static_cast<std::string>(cell.family_name()),
+        static_cast<std::string>(cell.column_qualifier()), cell.timestamp(),
+        static_cast<std::string>(cell.value())));
+  }
+  bigtable::BulkMutation bulk;
+  for (auto& kv : mutations) {
+    bulk.emplace_back(std::move(kv.second));
+  }
+  table.BulkApply(std::move(bulk));
+}
+
+int CellCompare(bigtable::Cell const& lhs, bigtable::Cell const& rhs) {
+  auto compare_row_key = lhs.row_key().compare(lhs.row_key());
+  if (compare_row_key != 0) {
+    return compare_row_key;
+  }
+  auto compare_family_name = lhs.family_name().compare(rhs.family_name());
+  if (compare_family_name != 0) {
+    return compare_family_name;
+  }
+  auto compare_column_qualifier =
+      lhs.column_qualifier().compare(rhs.column_qualifier());
+  if (compare_column_qualifier != 0) {
+    return compare_column_qualifier;
+  }
+  if (lhs.timestamp() < lhs.timestamp()) {
+    return -1;
+  }
+  if (lhs.timestamp() > lhs.timestamp()) {
+    return 1;
+  }
+  return lhs.value().compare(rhs.value());
+}
+
+bool CellLess(bigtable::Cell const& lhs, bigtable::Cell const& rhs) {
+  return CellCompare(lhs, rhs) < 0;
+}
+
+void PrintCells(std::ostream& os, std::vector<bigtable::Cell> const& cells) {
+  for (auto const& cell : cells) {
+    os << "  row_key=" << cell.row_key() << "\n  family=" << cell.family_name()
+       << "\n  column=" << cell.column_qualifier()
+       << "\n  timestamp=" << cell.timestamp() << "\n  value=" << cell.value();
+  }
+}
+
+void CheckEqual(absl::string_view where, std::vector<bigtable::Cell> expected,
```

Comment:
Would it be possible to add custom Matchers for Google Test instead? This is fine in the sense that it works, certainly, but I am just wondering if it would be better / more composable using Matchers, so we could say:

```cpp
EXPECT_THAT(actual, Equals(expected));
```

or similar. And maybe we just need to add a comparator for `bigtable::Cell`, and it would handle `std::vector<bigtable::Cell>` automatically for us?
 
Not suggesting you rewrite everything just because of this; just curious what you think.

Reply:
Certainly.  If/When (and it looks more like when now) we implement #153 that would be the way to go.  I think you have hinted enough, time to fix this.  Just one question before I do so:

Do you know if it is a really bad form to receive command-line arguments in a Google Test based program?  To avoid super-magical values, we need command-line arguments to figure out the project, instance, and table.  Also that makes the test usable with production Cloud Bigtable as well as the emulator.  Presumably we can parse the arguments and call `RUN_ALL_TESTS()` from our own `main()` function, but somebody will say something about non-hermetic-mumble-mumble-non-unit-mumble-tests.